### ISSUE TITLE
TSAN: [FBSDKLogger generateSerialNumber]

### ISSUE THREAD

**Issue by cltnschlosser** [facebook/facebook-ios-sdk#1693]
## Checklist

- [x] I've updated to the latest released version of the SDK
- [x] I've searched for existing [GitHub issues](https://github.com/facebook/facebook-ios-sdk/issues)
- [ ] I've looked for existing answers on [Stack Overflow](https://facebook.stackoverflow.com), the [Facebook Developer Community Forum](https://developers.facebook.com/community/) and the [Facebook Developers Group](https://www.facebook.com/groups/fbdevelopers)
- [x] I've read the [Code of Conduct](https://github.com/facebook/facebook-ios-sdk/blob/master/CODE_OF_CONDUCT.md)
- [x] This issue is not security related and can safely be disclosed publicly on GitHub

## Environment

Describe your dev environment here, giving as many details as possible. If you have them, make sure to include:

- Xcode Version: `12.4`
- Swift Version: `5.3.2` (if issue is Swift related)
- Installation Platform & Verison: `Cocoapods` version `1.10.1`

## Actual Results

```
file://<project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/Internal/FBSDKLogger.m: runtime: Threading Issues: Data race in +[FBSDKLogger generateSerialNumber] at g_serialNumberCounter

file://<project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/Internal/FBSDKLogger.m: notice: Threading Issues: 'g_serialNumberCounter' is a global variable (0x1121b9028)

Write of size 8 by thread 37
#0	0x000000010947f81b in +[FBSDKLogger generateSerialNumber] at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/Internal/FBSDKLogger.m:120
#1	0x0000000109459019 in -[FBSDKGraphRequestConnection logAndInvokeHandler:response:responseData:requestStartTime:] at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/GraphAPI/FBSDKGraphRequestConnection.m:995
#2	0x0000000109459b6a in -[FBSDKGraphRequestConnection taskDidCompleteWithResponse:data:requestStartTime:handler:] at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/GraphAPI/FBSDKGraphRequestConnection.m:1033
#3	0x0000000109449ff4 in __36-[FBSDKGraphRequestConnection start]_block_invoke_2 at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/GraphAPI/FBSDKGraphRequestConnection.m:253
#4	0x00007fff2351b6ca in ___lldb_unnamed_symbol142$$CFNetwork ()
#5	0x0000000121e4a9c8 in _dispatch_client_callout ()
Write of size 8 by thread 27
#0	0x000000010947f81b in +[FBSDKLogger generateSerialNumber] at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/Internal/FBSDKLogger.m:120
#1	0x0000000109459019 in -[FBSDKGraphRequestConnection logAndInvokeHandler:response:responseData:requestStartTime:] at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/GraphAPI/FBSDKGraphRequestConnection.m:995
#2	0x0000000109459b6a in -[FBSDKGraphRequestConnection taskDidCompleteWithResponse:data:requestStartTime:handler:] at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/GraphAPI/FBSDKGraphRequestConnection.m:1033
#3	0x0000000109449ff4 in __36-[FBSDKGraphRequestConnection start]_block_invoke_2 at <project_path>/Pods/FBSDKCoreKit/FBSDKCoreKit/FBSDKCoreKit/GraphAPI/FBSDKGraphRequestConnection.m:253
#4	0x00007fff2351b6ca in ___lldb_unnamed_symbol142$$CFNetwork ()
#5	0x0000000121e4a9c8 in _dispatch_client_callout ()
```

**Comment by joesus** [comment]
Thanks for reaching out to us with this issue! Unfortunately, there is not enough information included here to help debug the issue further. Specifically the SDK version or a sample app that can reproduce the issue would be helpful. As a result, we are closing this issue at this time. 

Please open a new issue using the Github issue template and be sure to include the requested information to help us debug the issue further OR report this to our designated developer support team here:  https://developers.facebook.com/support/bugs/

**Comment by cltnschlosser** [comment]
`g_serialNumberCounter` can be incremented from any thread and it has no protections. This was either with version 9.0.1 or 9.1.0. I recently upgraded.

**Comment by cltnschlosser** [comment]
https://github.com/facebook/facebook-ios-sdk/blob/d8c432aa15900bec062f9d099a1c05ec248aee20/FBSDKCoreKit/FBSDKCoreKit/Internal/FBSDKLogger.m#L120
This is a write that needs to be protected with a lock or some other serialization technique.

**Comment by joesus** [comment]
Yeah. Reopening. This is legitimate.

**Comment by joesus** [comment]
Should be fixed by c3b8b393e69ec576388e448029f2cedbfe1f42ee. Thanks for bringing this to our attention!
